docs/index.bs

```diff
@@ -2973,16 +2984,49 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                  Note: If the global object's associated list of event listeners does not have any event listener added at this moment, the service worker's [=set of event types to handle=] remains an empty set.
 
               1. Set |script|'s <a>has ever been evaluated flag</a>.
+              1. Unset the [=all fetch listeners are empty flag=].
+              1. If the [=Empty Fetch Handler Identification=] algorithm with |workerGlobalScope| returns true, the [=all fetch listeners are empty flag=] may be set.
+
+              Note: the user agents are encouraged to show warnings that the event listeners are no-op, and their executions can be skipped.
+
           1. Run the <a>responsible event loop</a> specified by |settingsObject| until it is destroyed.
           1. [=map/Clear=] |workerGlobalScope|'s [=map of active timers=].
       1. Wait for |serviceWorker| to be [=running=], or for |startFailed| to be true.
       1. If |startFailed| is true, then return *failure*.
       1. Return |serviceWorker|'s [=start status=].
   </section>
 
+  <section algorithm>
+    <h3 id="empty-fetch-handler-identification-algorithm"><dfn>Empty Fetch Handler Identification</dfn></h3>
+
+      : Input
+      :: |workerGlobalScope|, a [=service worker/global object=].
+      : Output
+      :: a boolean
+
+      1. If |workerGlobalScope|'s [=set of event types to handle=] does not [=set/contain=] <code>fetch</code>, then returns false.
+      1. Let |eventListenerList| be an empty [=list=].
+      1. [=list/For each=] |eventListener| of |workerGlobalScope|'s [=set of event types to handle=]:
+          1. If |eventListener|'s <a spec="dom">type</a> is <code>fetch</code>, then [=list/append=] |eventListener| to |eventListenerList|.
+      1. [=list/For each=] |eventListener| of |eventListenerList|:
+          1. If |eventListener|'s [=callback=] is not null:
+              1. Set |callback| to the result of [=converting|convert to an ECMAScript value=] |eventListener|'s [=callback=] to an ECMAScript value.
+              1. If [$IsCallable$](|callback|) is false:
+                  1. Let |getResult| be [=Completion=]([$Get$](|callback|), <code>handleEvent</code>).
+                  1. If |getResult| is [=abrupt completion=], then return false.
+                  1. Set |callback| to |getResult|.
+              1. If [$IsCallable$](|callback|) is false, then return false.
+              1. If |callback|'s [=function body=] is not empty (i.e. either a [=statement=] or [=declaration=] exist), then return false.
+
+              Note: it detects something like `onfetch = () => {}`. Some sites have a fetch event listener with empty body to make them recognized as progressive web application (PWA).
```

Comment:
@domenic I think this still doesn't work for `onfetch` as HTML wraps that function in its own function when creating the callback, right? See https://github.com/WebKit/standards-positions/issues/129#issuecomment-1431008116.

Reply:
Yes. We started discussing how to fix that (offline) but then got sidetracked into making the surrounding parts of the spec more precise.

To detech `onfetch = () => {}` I think the following would work, nested immediately under "If _eventListener_'s callback is not null":

> 1. If _eventListener_ is an `EventListener` of the type created by [activate an event handler](https://html.spec.whatwg.org/#activate-an-event-handler), then set |callback| to the result of [converting] _eventTarget_'s [event handler map](https://html.spec.whatwg.org/#event-handler-map)["`fetch`"]'s [value](https://html.spec.whatwg.org/#event-handler-value) to an ECMAScript value.
> 2. Otherwise, set |callback| to ...

(you could also factor out the [converting] into a separate step.)

Reply:
Can I ask where eventTarget in Step 1 come from?

Reply:
I am not fully understand the comment, but I tried to update the steps as suggested.
https://github.com/w3c/ServiceWorker/pull/1672/commits/27f0e006edf2af02c1c11a9421879aa11f467c51